Gradient Part 2 - Linear gradients

Dherse opened the pull request:
This is the second part of gradients, it includes the following changes:

- Adds the `gradient` type
- Adds the `gradient.linear` constructor for creating a gradient with the following args:
  - `stops` (variadic): either a color or a color and a ratio in an array
  - `dir` (defaults to `ltr`): the direction in which the ratio goes
  - `space`: the color space in which to interpolate the stops
  - `relative`: what is the gradient relative to (see later)
- Layout engine changes
  -  `Frame` can now be `FrameKind::Soft` or `FrameKind::Hard`, this is information being kept of simply to determine what is the "container" of an element
  - `block`, `box`, `columns`, `page`, `polygon`, and `frame`  produce hard frames, anything else produces a soft frame
  - hard frames are never inlined (we need to keep track of the transform)
- Adds the following method to the `gradient` type:
  - `stops` to get the color stops of the gradient
  - `space` to get the interpolation color space of the gradient
  - `relative` returns the "relative-ness" of a gradient (see later)
  - `dir` returns the angle of the gradient
  - `kind` returns the constructor used to create this gradient (currently only `gradient.linear`)
  - `sample` samples a color at a given ratio `t`
  - `samples` samples many colors (variadic) at a series of given ratios `t`
  - `sharp` turns the gradient into its sharp variant (see later)
  - `repeat` repeats the stops of the gradient `n` times, optionally mirror every other repetition
- Adds the following presets color maps (see later for an image with all of them): turbo, cividis, rainbow, spectral, viridis, inferno, magma, plasma, rocket, mako, vlag, icefire, flare, and crest. They can all be used (except rainbow) for data visualization.
- PDF/SVG/PNG export of gradients
- The following test:
  - aspect ratio correction (see later)
  - directions of gradients
  - all presets
  - `relative` argument
  - `repeat` method
  - `sharp` method
  - gradients are strokes
  - modified `methods.typ` to add tests to all methods in `gradient`
  - test that gradient on text (see later) gives an error
- Adds a `size + gradient` operation for creating a `stroke`
- Gradient on text causes an error with a hint saying it will be available soon:tm: (see later)

Also adds the following niceties:
- Adds better code completion for colors (was missing the new color spaces)
- Adds a `quadrant` method on `angle` that allows getting the quadrant in which the angle is, it is used when computing gradients, I figured it was a nice method to expose to the user.
- Small SVG export file size improvements

# Presets
![test](https://github.com/typst/typst/assets/9665250/7bbbb32a-f71f-4e5b-90ba-caea49c42f74)

# Gradients on text
Gradients on text are currently disabled, they will be included in **Part 4** of the gradient series of PRs. Currently, they give an error when someone tries to use them:
![image](https://github.com/typst/typst/assets/9665250/b5ab9f4e-04fb-431b-8e2c-f6b944b0dbc7)

# Relative-ness

Gradients can be relative to either the shape they are painted on, or to the nearest parent containers. This is controlled by the `relative` argument of the constructors. By default, gradients are relative to the shape they are painted on, unless they are painted on text, in which case they are relative to the parent.

The way the parent is determined is as follows:
- For shapes that are placed at the root/top level of the document, the parent is the page itself.
- For other shapes, the parent is the first  block or box that contains the shape. This includes the boxes and blocks that are implicitly created by show rules. For example, a rotate will not affect the parent of a gradient, but a grid will.

The following code exemplifies this, if they were not in the same relative coordinate space for their fills, the result would not look uniform and continuous.

```typ
#set page(width: 100pt, height: 100pt, margin: 0pt)
#set block(spacing: 0pt)

#let fill = gradient.linear(..gradient.icefire, relative: "parent", dir: ttb)

#place(top + left, square(fill: fill, size: 50pt))
#place(center + horizon, rotate(45deg, square(fill: fill, size: 50pt)))
#place(bottom + right, square(fill: fill, size: 50pt))
``` 

![test](https://github.com/typst/typst/assets/9665250/7f657537-c107-4f79-8f4b-2c3f0aea67c2)

# Sharp gradient

Additionally, the user can sample the gradient to make a "sharp" variant (i.e with sharp color bars), this is especially useful for creating a list of colors from a color map. Below a result of such a sharp gradient can be seen along with it's non-sharp version:

![test](https://github.com/typst/typst/assets/9665250/456a8b7f-a100-4eb7-a8e4-3c69e6dee333)

# Aspect ratio correction

Due to the way gradients are handled, they would normally ignore the aspect ratio of the shape and the `dir` of the gradient would become incorrect for gradients with a width to height ratio (aspect ratio) of anything but 1. This is handled in `angle.rs` with a small method. The reason why I mention it is because making this (imo necessary) correction leads to a larger file size as it decrease how well the exported gradient data can be re-used in PDF and SVG export.

PgBiel commented on crates/typst/src/doc.rs:
```diff
@@ -42,17 +42,48 @@ pub struct Frame {
+
+    /// Create a new, empty hard frame.
+    ///
+    /// Panics the size is not finite.
```
```suggestion
    /// Panics if the size is not finite.
```

PgBiel commented on crates/typst/src/doc.rs:
```diff
@@ -42,17 +42,48 @@ pub struct Frame {
+        }
+    }
+
+    /// Sets the frame's hardness
```
Based on previous style recommendations from Laurenz
```suggestion
    /// Sets the frame's hardness.
```

PgBiel commented on crates/typst/src/doc.rs:
```diff
@@ -407,6 +450,33 @@ impl Debug for Frame {
 
+#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
+pub enum FrameKind {
+    /// A container which follows its parent's size
+    Soft,
+
+    /// A container which uses its own size
+    Hard,
```
```suggestion
    /// A container which follows its parent's size.
    Soft,

    /// A container which uses its own size.
    Hard,
```

PgBiel commented on crates/typst/src/doc.rs:
```diff
@@ -407,6 +450,33 @@ impl Debug for Frame {
+    }
+}
+
+impl Default for FrameKind {
+    fn default() -> Self {
+        Self::Soft
+    }
+}
```
Might be worth it to document why this was chosen as the default

PgBiel commented on crates/typst/src/export/pdf/color.rs:
```diff
@@ -419,3 +427,78 @@ impl PaintEncode for Paint {
 }
+
+pub(super) trait CSFunctions {
+    /// Returns the range of the color space
+    fn range(self) -> [f32; 6];
+
+    /// Converts a color to the color space
```
```suggestion
    /// Returns the range of the color space.
    fn range(self) -> [f32; 6];

    /// Converts a color to the color space.
```

PgBiel commented on crates/typst/src/export/pdf/color.rs:
```diff
@@ -419,3 +427,78 @@ impl PaintEncode for Paint {
+        ]
+    }
+}
+
+pub(super) trait QuantizedColor {
+    fn quantize(color: f32, range: [f32; 2]) -> Self;
+}
```
Some brief description as docs could be useful here

PgBiel commented on crates/typst/src/export/pdf/gradient.rs:
```diff
@@ -0,0 +1,197 @@
+    let mut bounds = vec![];
+    let mut encode = vec![];
+
+    // Create the individual gradient functions for each pair of stops
+    for window in gradient.stops().windows(2) {
+        let (first, second) = (window[0], window[1]);
+
+        // Skip stops with the same position
```
```suggestion
    // Create the individual gradient functions for each pair of stops.
    for window in gradient.stops().windows(2) {
        let (first, second) = (window[0], window[1]);

        // Skip stops with the same position.
```

PgBiel commented on crates/typst/src/export/pdf/gradient.rs:
```diff
@@ -0,0 +1,197 @@
+        encode.extend([0.0, 1.0]);
+    }
+
+    // Special case for gradients with only two stops
+    if functions.len() == 1 {
+        return functions[0];
+    }
+
+    // Remove the last bound, since it's not needed for the stitching function
+    bounds.pop();
+
+    // Create the stitching function
```
```suggestion
    // Special case for gradients with only two stops.
    if functions.len() == 1 {
        return functions[0];
    }

    // Remove the last bound, since it's not needed for the stitching function.
    bounds.pop();

    // Create the stitching function.
```

PgBiel commented on crates/typst/src/export/pdf/page.rs:
```diff
@@ -222,6 +235,60 @@ struct State {
+    pub container_transform: Transform,
+    /// The size of the first hard frame in the hierarchy.
+    pub container_size: Size,
+    /// The size of the item
```
```suggestion
    /// The size of the item.
```

PgBiel commented on crates/typst/src/export/render.rs:
```diff
@@ -78,30 +78,83 @@ pub fn render_merged(
+        }
+    }
+
+    fn pre_translate(&self, pos: Point) -> Self {
```
not sure but might be a good idea to add some brief docs to these functions

PgBiel commented on crates/typst/src/export/render.rs:
```diff
@@ -626,12 +685,97 @@ impl From<Transform> for sk::Transform {
+                            .post_concat(state.transform.invert().unwrap()),
+                    }
+                });
+                let width =
+                    (container_size.x.to_f32() * state.pixel_per_pt).ceil() as u32;
+                let height =
+                    (container_size.y.to_f32() * state.pixel_per_pt).ceil() as u32;
```
Is there any risk of overflow here? (Should perhaps be documented, or handled)

Dherse replied:
If there is then it will happen (much) sooner in the rendering code, so it's a non issue for this specific case. That being said, it should be handled earlier in the chain since currently I'm pretty sure it will panic (should be another PR)

laurmaedje replied:
float -> int casts saturate as far as I know

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+/// A gradient is composed of a series of stops, each stop has a color and an offset.
+/// The offset is a [ratio]($ratio) between 0% and 100% that determines the position
+/// of the stop along the gradient. The color is the color of the gradient at that
+/// position. In Typst, you can chose to omit the offset, in which case it will be
```
```suggestion
/// position. In Typst, you can choose to omit the offset, in which case it will be
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+/// Currently gradients are not supported on text. However, in an upcoming release
+/// gradients will be supported on text.
+///
+/// ## Relative-ness
```
```suggestion
/// ## Relativeness
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+/// The way the parent is determined is as follows:
+/// - For shapes that are placed at the root/top level of the document, the parent
+///   is the page itself.
+/// - For other shapes, the parent is the first [`block`]($block) or [`box`]($box)
+///   that contains the shape. This includes the boxes and blocks that are implicitly
+///   created by show rules. For example, a [`rotate`]($rotate) will not affect the
+///   parent of a gradient, but a [`grid`]($grid) will.
```
I think the idea of "first" here is not very clear when thinking of the document as a tree structure. Perhaps it would be better to clarify whether it is "outermost" or "innermost" (I think those terms are clearer regarding who is the parent of whom).

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+/// Typst's gradients were designed to be widely compatible, however, in
+/// [PDF.js](https://mozilla.github.io/pdf.js/), the reader bundled with Firefox,
+/// gradients in `rotate` blocks may not be rendered correctly. This is a bug in
+/// PDF.js and not in Typst. Otherwise every type of gradient has been tested in
+/// every major PDF readers and browsers and should work as expected.
```
Added a "most" before browsers too since we just claimed that it might not work in firefox for some gradient types, but that's a debatable addition

```suggestion
/// PDF.js and not in Typst. Despite this, every type of gradient has been
/// tested in every major PDF reader, and should work in most browsers as
/// expected.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+///
+/// ## Presets
+///
+/// Typst also includes a number of preset color maps. In the following section the
+/// list of available presets is given, along with a sample of the gradient and
```
```suggestion
/// Typst also includes a number of preset color maps. In the following section, the
/// list of available presets is given, along with a sample of each gradient and
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+///
+/// The [`turbo`]($gradient.turbo) gradient is a rainbow-like gradient that is
+/// perceptually uniform. Turbo is a gradient that takes an optional number of
+/// stops, by default it is set to 20.
```
```suggestion
/// stops, which is set to 20 by default.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+/// The [`rainbow`]($gradient.rainbow) gradient is a rainbow gradient that is
+/// **not** perceptually uniform, therefore it should only be used for decorative
+/// purposes and not for data visualization. Rainbow is a gradient that takes an
+/// optional number of stops, by default it is set to 20. This gradient is best
```
```suggestion
/// optional number of stops, which is set to 20 by default. This gradient is best
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
````diff
@@ -0,0 +1,1070 @@
+/// #rect(width: 100pt, height: 20pt, fill: gradient.linear(..gradient.crest))
+/// ```
+///
+/// ### What about presets like "jet" and "parula"
````
```suggestion
/// ### Regarding presets like "jet" and "parula"
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+///
+/// - [Jet](https://jakevdp.github.io/blog/2014/10/16/how-bad-is-your-colormap/)
+///   is not a good color map, as it is not perceptually uniform. As such,
+///   it is not color blind friendly and should not be used for data visualization.
+///   And it is not included in Typst.
```
```suggestion
///   it is not color blind friendly and should not be used for data visualization,
///   due to which it is not included in Typst.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+///   is not a good color map, as it is not perceptually uniform. As such,
+///   it is not color blind friendly and should not be used for data visualization.
+///   And it is not included in Typst.
+/// - [Parula](https://www.mathworks.com/help/matlab/ref/parula.html)
+///   is a good color map included in matlab, but it is not included as
+///   a preset in Typst. This is because it is owned by MathWorks and is not public.
```
```suggestion
/// - [Parula](https://www.mathworks.com/help/matlab/ref/parula.html)
///   is a good color map included in MATLAB, but it is not included as
///   a preset in Typst. This is because it is owned by MathWorks and is not public.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+        /// - `"parent"`: The gradient is relative to the bounding box of the
+        ///   parent that contains the element onto which the gradient is applied.
+        ///
+        /// For an element placed at the root/top level of the document, the parent
+        /// is the page itself. For other elements, the parent is the first block or
+        /// box that contains the element.
```
Same comment regarding the usage of the word "first"

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+        }))
+    }
+
+    /// Return the stops of this gradient.
```
For consistency with the other functions
```suggestion
    /// Returns the stops of this gradient.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+    pub fn sharp(
+        &self,
+
+        /// The number of stops in the gradient.
+        n: Spanned<usize>,
```
Perhaps rename this variable to `stops`, or `stop_count`?

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+            .collect::<Vec<_>>();
+
+        let mut positions = Vec::with_capacity(n.v * 2);
+        let p = |i| i as f64 * 1.0 / n.v as f64;
+
+        let t = smoothness * 1.0 / (4.0 * n.v as f64);
+        for i in 0..n.v {
+            let mut j = 2 * i;
+            positions.push(p(i));
+            if j > 0 {
+                positions[j] += t;
+            }
+
+            j += 1;
+            positions.push(p(i + 1));
+            if j < colors.len() - 1 {
+                positions[j] -= t;
+            }
```
Perhaps those variables could have a bit more meaningful names? (I know this is mostly maths:tm:, but maybe we can turn these into names which are a bit more helpful to future maintainers, or maybe document what they mean?)

Dherse replied:
Truth is, I wrote this code months ago and even I struggle to understand it, as far as `p`, `t`, `i`, and `j`:
- `p` and `t` I can rename
- `i` and `j` are very common variables for indices

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+    pub fn repeat(
+        &self,
+
+        /// The number of times to repeat the gradient.
+        n: Spanned<usize>,
```
Perhaps `repeats` or `repeat_count` instead of `n`?

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+
+    /// Creates a [turbo] stop list.
+    ///
+    /// You can control the number of stops in the gradient using the `stops` parameter, by default it is set to 20.
```
```suggestion
    /// You can control the number of stops in the gradient using the `stops`
    /// parameter, which is set to 20 by default.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+
+    /// Creates a [cividis] stop list.
+    ///
+    /// You can control the number of stops in the gradient using the `stops` parameter, by default it is set to 20.
```
```suggestion
    /// You can control the number of stops in the gradient using the `stops`
    /// parameter, which is set to 20 by default.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+    /// Creates a rainbow list of color stops with the given parameters.
+    ///
+    /// You can control the number of stops in the gradient using the `stops`
+    /// parameter, by default it is set to 20.
```
```suggestion
    /// parameter, which is set to 20 by default.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+        .into_value())
+    }
+
+    /// Creates a rainbow list of color stops with the given parameters.
```
Perhaps this is clearer wording?
```suggestion
    /// Creates a list of rainbow color stops with the given parameters.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+    pub fn sample_at(&self, (x, y): (f32, f32), (width, height): (f32, f32)) -> Color {
+        let t = match self {
+            Self::Linear(linear) => {
+                // normalize the coordinates
+                let (mut x, mut y) = (x / width, y / height);
+
+                // Handle the direction of the gradient
+                let angle = linear.angle.to_rad().rem_euclid(TAU);
+
+                // Aspect ratio correction
```
```suggestion
                // Normalize the coordinates.
                let (mut x, mut y) = (x / width, y / height);

                // Handle the direction of the gradient.
                let angle = linear.angle.to_rad().rem_euclid(TAU);

                // Aspect ratio correction.
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+
+#[derive(Clone, Hash, PartialEq, Eq)]
+pub struct LinearGradient {
+    /// The color stops of this gradient
+    pub stops: Vec<(Color, Ratio)>,
+
+    /// The direction of this gradient
+    pub angle: Angle,
+
+    /// The color space in which to interpolate the gradient
+    pub space: ColorSpace,
+
+    /// The relative placement of the gradient
+    pub relative: Smart<Relative>,
+
+    /// Whether to anti-alias the gradient (used for sharp gradient)
```
```suggestion
    /// The color stops of this gradient.
    pub stops: Vec<(Color, Ratio)>,

    /// The direction of this gradient.
    pub angle: Angle,

    /// The color space in which to interpolate the gradient.
    pub space: ColorSpace,

    /// The relative placement of the gradient.
    pub relative: Smart<Relative>,

    /// Whether to anti-alias the gradient (used for sharp gradients).
```

PgBiel commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1070 @@
+    }
+}
+
+/// What is the gradient relative to:
+/// - Itself (its own bounding box)
+/// - Its parent (the parent's bounding box)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Relative {
+    This,
+    Parent,
+}
```
Should produce more useful documentation hints when coding

```suggestion
/// What is the gradient relative to.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Relative {
    /// The gradient is relative to itself (its own bounding box).
    This,
    
    /// The gradient is relative to its parent (the parent's bounding box).
    Parent,
}
```

PgBiel reviewed (commented):
Nitpick moment

Dherse commented:
While doing my own "code review", I found a bug with strokes (specifically lines), I fixed it for all three export targets, while at it I also figured a pretty big file optimization for SVG export, and I added tests to make sure this issue is well and truly fixed as well as testing it in PDF export manually and it all works 🎉

reknih commented on tests/typ/visualize/gradient-relative.typ:
```diff
@@ -0,0 +1,34 @@
+---
+
+#let grad = gradient.linear(red, blue, green, purple, relative: "parent");
+#let my_rect = rect(width: 50%, height: 50%, fill: grad)
```
Canonical Typst style is `kebab-case`

Dherse replied:
Oh right sorry

reknih commented on tests/typ/visualize/gradient-relative.typ:
```diff
@@ -0,0 +1,34 @@
+---
+
+#let grad = gradient.linear(red, blue, green, purple, relative: "self");
+#let my_rect = rect(width: 50%, height: 50%, fill: grad)
```
Same here

reknih commented on tests/typ/visualize/gradient-transform.typ:
```diff
@@ -0,0 +1,16 @@
+---
+
+#let grad = gradient.linear(red, blue, green, purple, relative: "parent");
+#let my_rect = rect(width: 50pt, height: 50pt, fill: grad)
```
Same here

reknih commented on tests/typ/visualize/gradient-transform.typ:
```diff
@@ -0,0 +1,16 @@
+// Test whether `relative: "parent"` works correctly
+// In the first test, the image should look as if there is a single
+// gradient that is being used for both the page and the rectangles.
+// In the second test, the image should look as if there are multiple
+// gradients, one for each rectangle.
```
Same comment as in `gradient-relative.typ`, remove the comment or add a unique one.

Dherse replied:
Copy pasting error, I’ll correct once I’m home ;)

reknih commented on crates/typst/src/geom/transform.rs:
```diff
@@ -68,6 +68,68 @@ impl Transform {
+        }
+
+        let det = self.sx * self.sy - self.kx * self.ky;
+        if det.get() < 1e-12 {
+            panic!("Transform is not invertible");
+        }
```
Is it truly impossible for transformations in Typst to be irreversible?

Dherse replied:
I’ll double check, but I think not, it would require the scale to be zero, something I could just make an error into `transform` and `scale` functions, basically preventing the condition earlier in the chain (since it’s required for PNG rendering too and it’s only in cases that are basically nonsensical for typst anyways.

laurmaedje replied:
This should return `Option<Self>` in my opinion. If at all, it's better to unwrap at a call site that has more context information.

reknih commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1088 @@
+///
+/// A gradient is composed of a series of stops, each stop has a color and an offset.
+/// The offset is a [ratio]($ratio) between 0% and 100% that determines the position
+/// of the stop along the gradient. The color is the color of the gradient at that
```
```suggestion
/// of the stop along the gradient. The stop's color is the color of the gradient at that
```

A bit awkward otherwise

reknih commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1088 @@
+/// ## Color spaces and interpolation
+///
+/// Gradients can be interpolated in any color space. By default, gradients are
+/// interpolated in the [Oklab]($color.oklab) color space, which is a perceptually
```
Add a link to an appropriate source if the reader wants to learn more?

reknih commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1088 @@
+/// ## Note on compatibility
+///
+/// Typst's gradients were designed to be widely compatible; however, in
+/// [PDF.js](https://mozilla.github.io/pdf.js/), the reader bundled with Firefox,
```
```suggestion
/// [PDF.js](https://mozilla.github.io/pdf.js/), the PDF reader bundled with Firefox,
```

reknih commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1088 @@
+///
+/// ### Rainbow
+///
+/// The [`rainbow`]($gradient.rainbow) gradient is a rainbow gradient that is
+/// **not** perceptually uniform, therefore it should only be used for decorative
+/// purposes and not for data visualization. Rainbow is a gradient that takes an
+/// optional number of stops, which is set to 20 by default. This color map is best
+/// used by setting the interpolation color space to [HSL]($color.hsl).
+///
+/// ❌ This color map is **not** suitable for data visualization.
```
I would remove the ✅ ... messages everywhere and instead write a notice like this here:

```suggestion
/// The [`rainbow`]($gradient.rainbow) gradient cycles through the full
/// color spectrum. Rainbow is a gradient that takes an
/// optional number of stops, which is set to 20 by default. This color map is best
/// used by setting the interpolation color space to [HSL]($color.hsl).
///
/// **Attention:** The rainbow gradient is **not** suitable for data visualization. It is
/// **not** perceptually uniform, so the differences between values become unclear to your readers.
/// It should only be used for decorative purposes. 
```

Ignore the line breaks.

reknih commented on crates/typst/src/geom/gradient.rs:
````diff
@@ -0,0 +1,1088 @@
+/// #rect(width: 100pt, height: 20pt, fill: gradient.linear(..gradient.crest))
+/// ```
+///
+/// ### Regarding presets like "jet" and "parula"
+///
+/// - [Jet](https://jakevdp.github.io/blog/2014/10/16/how-bad-is-your-colormap/)
+///   is not a good color map, as it is not perceptually uniform. As such,
+///   it is not color blind friendly and should not be used for data visualization,
+///   due to which it is not included in Typst.
+/// - [Parula](https://www.mathworks.com/help/matlab/ref/parula.html)
+///   is a good color map included in MATLAB, but it is not included as
+///   a preset in Typst. This is because it is owned by MathWorks and is not public.
````
Do we need these?

Dherse replied:
Judging by the number of people asking this on the different com channels for matplotlib, I am afraid so :(

Dherse commented on tests/ref/visualize/gradient-aspect-ratio.png:
An artefact, most likely float precision issues, nothing that can really be done here, they’re not present in PDF or SVG export.

laurmaedje commented on crates/typst/src/export/svg.rs:
```diff
@@ -58,6 +62,73 @@ struct SVGRenderer {
+    /// These are the actual gradients being written in the SVG file.
+    /// These gradients are deduplicated because they do not contain the transform
+    /// matrix, allowing them to be reused across multiple invocations.
+    gradients: Deduplicator<(Gradient, Ratio)>,
```
What is the `Ratio`?

Dherse replied:
Added doc, basically it's the aspect ratio to correct the angle.

laurmaedje commented on crates/typst/src/export/svg.rs:
```diff
@@ -100,25 +173,36 @@ impl SVGRenderer {
         }
 
         for (pos, item) in frame.items() {
+            // File size optimization
```
This is good for now, but ideally we wouldn't wrap every single frame item in a `<g>`.

Dherse replied:
I agree, it makes debugging SVGs a lot harder since there are so many nested level, but I don't want to accidentally break it all of now so I just added small file size optimizations where I stumbled on them.

laurmaedje commented on crates/typst/src/export/svg.rs:
```diff
@@ -270,42 +362,149 @@ impl SVGRenderer {
+        }
+    }
+
+    /// Pushes a gradient to the SVG file. If the gradient is already present,
```
It's not really pushed to `SVG file` right? More to an intermediate structure and it will be written later.

laurmaedje commented on crates/typst/src/export/svg.rs:
```diff
@@ -408,6 +617,107 @@ impl SVGRenderer {
+                            let t0 = i as f64 / 32.0;
+                            let t = start_t + (end_t - start_t) * t0;
+                            let c = Color::mix(
+                                vec![
+                                    WeightedColor::new(start_c, 1.0 - t0),
+                                    WeightedColor::new(end_c, t0),
+                                ],
```
This allocates potentially quite a lot. Either we have a separate mix function with slices or we could at least preallocate and reuse the same vector.

laurmaedje commented on crates/typst/src/geom/angle.rs:
```diff
@@ -66,10 +68,32 @@ impl Angle {
         self.to_rad().cos()
     }
 
+    /// Get the sine and cosine of this angle in radians.
+    pub fn sin_cos(self) -> (f64, f64) {
+        self.to_rad().sin_cos()
+    }
```
The std impl is literally `(self.sin(), self.cos())`, so it doesn't seem to be a perf thing. Which makes me wonder whether this is really worth the few characters less. It seems like a somewhat strange abstraction to me.

Dherse replied:
I agree, my best guess is that it helps the compiler optimize better on some targets? I like the slightly shorter nature, but it doesn't do much apart from looking slightly nicer.

laurmaedje replied:
I would remove it.

laurmaedje commented on crates/typst/src/geom/angle.rs:
```diff
@@ -85,6 +109,27 @@ impl Angle {
         self.to_unit(AngleUnit::Deg)
     }
+
+    /// Get the quadrant of the Cartesian plane that this angle lies in.
+    /// The angle is automatically normalized to the range `0deg..=360deg`.
```
```suggestion
    /// Get the quadrant of the Cartesian plane that this angle lies in.
    ///
    /// The angle is automatically normalized to the range `0deg..=360deg`.
```

laurmaedje commented on crates/typst/src/geom/angle.rs:
```diff
@@ -66,10 +68,32 @@ impl Angle {
     }
+
+    /// Corrects this angle for the aspect ratio of a gradient.
+    pub fn correct_aspect_ratio(self, aspect_ratio: Ratio) -> Self {
```
should this function really live with the `Angle` type? It feels like it belongs with gradients.

laurmaedje commented on crates/typst/src/geom/angle.rs:
```diff
@@ -85,6 +109,27 @@ impl Angle {
+    /// - Fourth: `270deg..=360deg` (bottom-right)
+    #[func]
+    pub fn quadrant(self) -> Quadrant {
+        let rad = self.to_rad().rem_euclid(TAU);
+        if (0.0..=FRAC_PI_2).contains(&rad) {
+            Quadrant::First
+        } else if (FRAC_PI_2..=PI).contains(&rad) {
+            Quadrant::Second
+        } else if (PI..=(3.0 * FRAC_PI_2)).contains(&rad) {
+            Quadrant::Third
+        } else {
+            Quadrant::Fourth
+        }
```
would this logic perhaps be simpler in `to_deg()`?

Dherse replied:
I mean same same, I just write it in a less convoluted way ;)

laurmaedje commented on crates/typst/src/geom/angle.rs:
```diff
@@ -192,6 +237,53 @@ impl Debug for AngleUnit {
+        2 => Self::Second,
+        3 => Self::Third,
+        4 => Self::Fourth,
+        _ => bail!("Invalid quadrant value: {}", v),
```
```suggestion
        _ => bail!("quadrant must be 1, 2, 3, or 4, but is {v}"),
```

laurmaedje commented on crates/typst/src/geom/paint.rs:
```diff
@@ -5,6 +5,21 @@ use super::*;
+
+impl Paint {
+    // TODO: implement gradients on text.
+    /// Temporary method to unwrap a solid color used
+    /// for text rendering.
```
```suggestion
    /// Temporary method to unwrap a solid color used for text rendering.
```

laurmaedje commented on crates/typst/src/geom/path.rs:
```diff
@@ -51,4 +51,28 @@ impl Path {
     }
+
+    /// Computes the size of the path.
+    pub fn size(&self) -> Size {
```
`size` doesn't feel super well-defined. maybe `bounds_size` / `bbox_size` or something? just `bounds` / `bbox` might be confusing because it doesn't return a rect, only a size.

laurmaedje commented on crates/typst/src/geom/path.rs:
```diff
@@ -51,4 +51,28 @@ impl Path {
+        let mut max_x = -Abs::inf();
+        let mut max_y = -Abs::inf();
+
+        for pt in self.0.iter() {
+            match pt {
+                PathItem::MoveTo(pt)
+                | PathItem::LineTo(pt)
+                | PathItem::CubicTo(_, _, pt) => {
+                    min_x = min_x.min(pt.x);
+                    min_y = min_y.min(pt.y);
+                    max_x = max_x.max(pt.x);
+                    max_y = max_y.max(pt.y);
+                }
+                PathItem::ClosePath => (),
+            }
+        }
```
```suggestion
        for item in self.0.iter() {
            match item {
                PathItem::MoveTo(pos)
                | PathItem::LineTo(pos)
                | PathItem::CubicTo(_, _, pos) => {
                    min_x = min_x.min(pos.x);
                    min_y = min_y.min(pos.y);
                    max_x = max_x.max(pos.x);
                    max_y = max_y.max(pos.y);
                }
                PathItem::ClosePath => (),
            }
        }
```
`pt` sounds like the unit.

also: does this really compute the correct bounding box? A cubic bezier curve isn't necessarily bounded by it's two end points.

Dherse replied:
Hmmm, that's very true, I'll quickly fix it

laurmaedje commented on crates/typst/src/geom/ratio.rs:
```diff
@@ -134,6 +134,12 @@ impl Div<Ratio> for f64 {
     }
 }
 
+impl From<Abs> for Ratio {
+    fn from(value: Abs) -> Self {
+        Self::new(value.to_pt())
+    }
+}
```
This impl feels _very_ wrong. What is it used for?

Dherse replied:
I mean I don't see what's so wrong about it, but it *was* used to make some `Transform::scale` a bit less of a chore to write, but it's okay it's removed

laurmaedje replied:
It's wrong in my opinion because there is no natural conversion from a length to a ratio. Why should 50pt coerce into 50%? It doesn't make sense in isolation.

laurmaedje commented on crates/typst/src/geom/shape.rs:
```diff
@@ -32,4 +32,13 @@ impl Geometry {
     }
+
+    /// The size of the geometry.
+    pub fn size(&self) -> Size {
```
related to the path function if something would change

laurmaedje commented on crates/typst/src/geom/size.rs:
```diff
@@ -18,6 +18,11 @@ impl Size {
         Point::new(self.x, self.y)
     }
+
+    /// Convert to a ratio.
```
The doc comment could be clearer since there isn't a natural conversion from size to ratio.

laurmaedje commented on crates/typst/src/geom/transform.rs:
```diff
@@ -68,6 +68,68 @@ impl Transform {
+        }
+    }
+
+    /// Convert to an array of floats.
```
I feel like this should live with the PDF exporting code because it assumes the unit as points. It doesn't make sense in isolation.

laurmaedje commented on tests/ref/compiler/repr.png:
I think the color and gradient tests should be extracted and rendered onto a larger page. The rendering is getting pretty messy here.

laurmaedje commented on tests/typ/visualize/gradient-aspect-ratio.typ:
```diff
@@ -0,0 +1,22 @@
+#set page(height: 120pt, width: 110pt, margin: 0pt)
+#set block(spacing: 0pt)
+
+#for i in range(0, 360, step: 30) {
```
could you apply 2 space indent everywhere and add newlines to the end of all files?

laurmaedje commented on tests/ref/visualize/gradient-aspect-ratio.png:
Most other tests are clear, but I don't really follow what is going on here.

Dherse replied:
All it does is checking that the angle is the same irrespective of the aspect ratio I'll change it to be clearer

laurmaedje commented on tests/ref/visualize/gradient-presets.png:
I think it would be nice if this one would look a bit nicer, with a bit more spacing and room to breath and centered text. Since it's the demo™ gradient test.

Dherse replied:
Done

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+/// details on the progress of gradients.
+///
+/// ## Stops
+///
```
Could you remove the newlines below headings?

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+///   ("HSV", color.hsv),
+/// )
+///
+/// #for space in spaces {
+///   block(
+///     width: 100%,
+///     height: 10pt,
+///     fill: gradient.linear(red, blue, space: space.at(1))
+///   )[
+///     #space.at(0)
+///   ]
+/// }
```
```suggestion
/// #for (name, space) in spaces {
///   block(
///     width: 100%,
///     height: 10pt,
///     fill: gradient.linear(red, blue, space: space),
///     name,
///   )
/// }
```

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+///
+/// ## Direction
+///
+/// Some gradients are sensitive to the direction of the gradient. For example, a
+/// linear gradient has an angle that determines the direction of the gradient. Instead
```
the word gradient is a bit overused in these two sentences.

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+///
+/// Some gradients are sensitive to the direction of the gradient. For example, a
+/// linear gradient has an angle that determines the direction of the gradient. Instead
+/// of the traditional clockwise angle, Typst uses an anti-clockwise angle, with 0° being
```
why do we go against the tradition?

Dherse replied:
Because we're not and that snippet of the docs is wrong?

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+///
+/// ## Presets
+///
+/// Typst also includes a number of preset color maps. In the following section, the
```
Is "color maps" an appropriate term for gradients?

Dherse replied:
I designed the gradient PR to be able to specifically do both jobs (for plotting packages) so I would say so, and since those are not gradients, the correct term is color map.

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+/// - [Parula](https://www.mathworks.com/help/matlab/ref/parula.html)
+///   is a good color map included in MATLAB, but it is not included as
+///   a preset in Typst. This is because it is owned by MathWorks and is not public.
+///
```
```suggestion
```

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+}
+
+impl Debug for Gradient {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
```suggestion
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
```
with import `use std::fmt::{self, Debug, Formatter};`

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+
+        /// The relative placement of the gradient.
+        ///
+        /// - `"this"`: The gradient is relative to the bounding box of the
+        ///   container onto which it is painted.
+        /// - `"parent"`: The gradient is relative to the bounding box of the
+        ///   parent that contains the element onto which the gradient is applied.
```
If you derive `Cast` on `Relative` and add the comments as doc comments of the enum variants, the docs would automatically render this correctly, so that you don't need to add this list here.

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+        ///   parent that contains the element onto which the gradient is applied.
+        ///
+        /// For an element placed at the root/top level of the document, the parent
+        /// is the page itself. For other elements, the parent is the innermost block
+        /// or box that contains the element.
```
aren't there other hard frames?

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+        }))
+    }
+
+    /// Returns the stops of this gradient.
```
will all the accessor methods below make sense for all gradient kinds down the road? in particular, things like `angle` or `dir`?

Dherse replied:
Generally yes, `dir` will just return zero (or could be none) for a radial gradient, otherwise they all make sense

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+impl Gradient {
+    /// Samples the gradient at a given position, in the given container.
+    /// Handles the aspect ratio and angle directly.
+    pub fn sample_at(&self, (x, y): (f32, f32), (width, height): (f32, f32)) -> Color {
```
wouldn't it make more sense for this to take normalize coordinates in the first place?

Dherse replied:
No because of the aspect ratio compensation, or it would just take the aspect ratio as a separate param which is basically the same

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+cast! {
+    Stop,
+    self => if let Some(offset) = self.offset {
+        array![ self.color.into_value(), offset ].into_value()
```
```suggestion
        array![self.color.into_value(), offset].into_value()
```

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+    angle: Angle => Self::Angle(angle),
+}
+
+impl From<RatioOrAngle> for f64 {
+    fn from(value: RatioOrAngle) -> Self {
+        match value {
+            RatioOrAngle::Ratio(ratio) => ratio.get(),
+            RatioOrAngle::Angle(angle) => angle.to_rad().rem_euclid(TAU) / TAU,
+        }
+        .clamp(0.0, 1.0)
+    }
+}
```
I think this should be a method on `RatioOrAngle` instead.

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+                    *span,
+                    "either all stops must have an offset or none of them can"
+                )
+                .with_hint("try adding an offset to all stops"));
```
very nice work on the hints :)

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+            };
+
+            if stop.get() < last_stop {
+                bail!(*span, "offsets must be in stricly monotonic");
```
```suggestion
                bail!(*span, "offsets must be in stricly monotonic order");
```

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+    .unwrap()
+}
+
+fn cubehelix_to_rgb(h: f32, s: f32, l: f32) -> Color {
```
Maybe this could be moved into `rainbow`? (or even inlined there)

Dherse replied:
It's true that there used to be more cubehelix presets, I'll move it

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+
+macro_rules! preset {
+    ($name:ident; $($colors:literal),* $(,)*) => {
+        #[comemo::memoize]
```
A once cell `Lazy` would probably be faster. Probably not particular performance-sensitive though.

Dherse replied:
I did the change anyway :)

laurmaedje commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+
+#[scope]
+impl Gradient {
+    pub const SPECTRAL: fn() -> Array = spectral;
+    pub const VIRIDIS: fn() -> Array = viridis;
+    pub const INFERNO: fn() -> Array = inferno;
+    pub const MAGMA: fn() -> Array = magma;
+    pub const PLASMA: fn() -> Array = plasma;
+    pub const ROCKET: fn() -> Array = rocket;
+    pub const MAKO: fn() -> Array = mako;
+    pub const VLAG: fn() -> Array = vlag;
+    pub const ICEFIRE: fn() -> Array = icefire;
+    pub const FLARE: fn() -> Array = flare;
+    pub const CREST: fn() -> Array = crest;
```
It feels wrong to me that the presets are _arrays,_ not gradients. I guess it has to be that way because of other configuration and gradient kind etc., but still wanted to mention it.

Dherse replied:
I agree with you, I thought of making them sort of constructors of their own, but it would have lead to so much duplication. I am open to any ideas regarding that!

PgBiel replied:
Perhaps the gradient constructors could accept other gradients as a parameter as well, automatically inheriting their properties? That way, the presets would have some good defaults, and would be easily customizable with `gradient.linear(rainbow, something_else, ...)`. (Another possibility would be to only inherit the stops from the gradient, but I'm not sure if that could be considered inconsistent.)

Dherse replied:
That simply does not work as there is no « default » gradient and once the other gradients kinds are added that won’t make much sense. One solution could be to add a `colormap` type and make them as such (on which you can call `.stops()` and which can be taken directly as an argument, but otherwise it’s not doable in a nicer way than an array.

Dherse commented on crates/typst-ide/src/complete.rs:
```diff
@@ -1169,6 +1169,27 @@ impl<'a> CompletionContext<'a> {
+                    );
+                    self.snippet_completion(
+                        "color.linear-rgb()",
+                        "linear-rgb(${r}, ${g}, ${b}, ${a})",
```
indeed, on linear-rgb, fixed

Dherse commented on crates/typst-library/src/layout/transform.rs:
```diff
@@ -117,6 +117,7 @@ impl Layout for RotateElem {
             .pre_concat(Transform::rotate(self.angle(styles)))
             .pre_concat(Transform::translate(-x, -y));
         frame.transform(ts);
+        frame.set_kind(FrameKind::Soft);
```
I explicitly set `transform` frames as soft, it does not break any hard frame that would be in it

laurmaedje replied:
I don't completely follow. The frame returned by `self.body().layout(vt, styles, pod)?.into_frame()` might be a hard frame from a block or something. If you set it to soft, it's not hard anymore, right? Or am I missing something?

Dherse replied:
Should I just wrap it into a soft frame by default? Or just not touch it at all perhaps?

Dherse replied:
I just removed the `set_kind`  and it works as expected, perhaps I was just being overly zealous with my use of soft frames

Dherse commented on crates/typst/src/doc.rs:
```diff
@@ -398,6 +430,17 @@ impl Frame {
+/// Tools for rendering.
+impl Frame {
+    /// The size the frame should be considered for scaling gradients.
+    pub fn render_size(&self, parent_size: Size) -> Size {
```
I removed it, I used it in my old impl, but since cleaning up it's not needed any more ;)

Dherse commented on crates/typst/src/doc.rs:
```diff
@@ -407,6 +450,35 @@ impl Debug for Frame {
+    }
+}
+
+impl Default for FrameKind {
```
I kept soft as a default but with a better justification. I chose `Soft` because it's the most common frame type, but beyond that, it's also the type that does not "touch" the hierarchy for gradients. Therefore imo it's the best preset.

Dherse commented on crates/typst/src/export/pdf/color.rs:
```diff
@@ -302,17 +302,21 @@ impl ColorEncode for ColorSpace {
-    fn set_as_fill(&self, ctx: &mut PageContext) {
-        let Paint::Solid(color) = self;
+    fn set_as_fill(&self, ctx: &mut PageContext, transforms: Transforms) {
+        let color = match self {
```
I did one better, I added an impl to `Color` (just like I did for `Gradient`) and made the `Paint` impl just delegate to this, it's the cleanest solution imo

Dherse commented on crates/typst/src/export/pdf/color.rs:
```diff
@@ -419,3 +427,80 @@ impl PaintEncode for Paint {
+    }
+}
+
+impl QuantizedColor for u16 {
+    fn quantize(color: f32, range: [f32; 2]) -> Self {
+        let value = (color - range[0]) / (range[1] - range[0]);
+        (value.max(0.0).min(1.0) * Self::MAX as f32)
+            .round()
+            .max(0.0)
+            .min(Self::MAX as f32) as Self
+    }
+}
+
+impl QuantizedColor for u32 {
+    fn quantize(color: f32, range: [f32; 2]) -> Self {
+        let value = (color - range[0]) / (range[1] - range[0]);
+        (value.max(0.0).min(1.0) * Self::MAX as f32)
+            .round()
+            .max(0.0)
+            .min(Self::MAX as f32) as Self
+    }
+}
```
I'll start by removing the ones I know won't be used

Dherse replied:
That removed all of the duplication ;)

Dherse commented on crates/typst/src/export/pdf/color.rs:
```diff
@@ -419,3 +427,80 @@ impl PaintEncode for Paint {
+        let range = self.range();
+        let [x, y, z, _] = color.to_space(self).to_vec4();
+
+        // We need to add 0.4 to y and z for Oklab
+        // This is because DeviceN color spaces in PDF can **only** be in
+        // the range 0..1 and some readers enforce that.
+        // Also map the angle range of HSV to 0..1
```
Added more doc ;)

Dherse commented on crates/typst/src/export/pdf/gradient.rs:
```diff
@@ -0,0 +1,199 @@
+        gradient: gradient.clone(),
+    };
+
+    let index = ctx.parent.gradient_map.insert(pdf_gradient);
```
I am unsure how else I would solve this?

laurmaedje replied:
It's probably fine.

Dherse commented on crates/typst/src/export/pdf/page.rs:
```diff
@@ -287,9 +368,15 @@ impl PageContext<'_, '_> {
+
+    fn set_fill(&mut self, fill: &Paint, transforms: Transforms) {
+        if self.state.fill.as_ref() != Some(fill)
+            || matches!(self.state.fill, Some(Paint::Gradient(_)))
```
That's exactly why and unfortunately for PDF there is no alternative, I can deduplicate some things which I am not doing but plan to in Part 6: #2282 but for now it's a good implementation, despite what it looks like gradients (especially linear ones) are fairly conservative in terms of the space they take so it's not that big a deal

Dherse commented on crates/typst/src/export/render.rs:
```diff
@@ -78,30 +78,88 @@ pub fn render_merged(
+    container_transform: sk::Transform,
+
+    /// The mask of the current item.
+    mask: Option<Arc<sk::Mask>>,
```
You're right, I didn't want to have to annotate a lifetime, but it was pretty painless in the end ;)

Dherse commented on crates/typst/src/export/svg.rs:
```diff
@@ -100,25 +173,36 @@ impl SVGRenderer {
 
     /// Render a frame with the given transform.
-    fn render_frame(&mut self, frame: &Frame, ts: Transform) {
+    fn render_frame(&mut self, state: State, transform: Transform, frame: &Frame) {
```
Apart from me not liking it? no. In this case, I had just gone back over after removing the `ts` arg, and then re-adding it later, whoops

bluebear94 commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+///
+/// - [Jet](https://jakevdp.github.io/blog/2014/10/16/how-bad-is-your-colormap/)
+///   is not a good color map, as it is not perceptually uniform. As such,
+///   it is not color blind friendly and should not be used for data visualization,
+///   due to which it is not included in Typst.
```
```suggestion
///   it is not color blind friendly and should not be used for data visualization;
///   thus, it is not included in Typst.
```

PgBiel replied:
well, both are correct
but I'm ok with changing it

Dherse replied:
If both are correct then I see no reason to change and drag this PR down any more.

PgBiel commented on tests/typ/compiler/repr-color-gradient.typ:
```diff
@@ -0,0 +1,22 @@
+#gradient.linear(blue, red) \
+#gradient.linear(blue, red, dir: 45deg) \
+#gradient.linear(blue, red, dir: 45deg, relative: "self") \
+#gradient.linear(blue, red, dir: 45deg, space: rgb)
```
plz add newlines at the end of the test files :smiley:

Dherse replied:
This is not something that is consistent across the code base, hence why I didn’t, if you want to go ahead but I have spent way more than enough time on this as it is to get bogged down in minute, nonsensical, nitpicks

Dherse replied:
Sorry, I had just driven for two hours, I was in quite a mood.

And I thought of this: `cargo fmt` strips the last line, shouldn't it be the same for tests?

laurmaedje replied:
I don't `cargo fmt` does that? At least not in my setup. It is a bit inconsistent across the code base because I don't bring it up anymore in PRs.

bluebear94 commented on crates/typst/src/geom/gradient.rs:
```diff
@@ -0,0 +1,1089 @@
+///   due to which it is not included in Typst.
+/// - [Parula](https://www.mathworks.com/help/matlab/ref/parula.html)
+///   is a good color map included in MATLAB, but it is not included as
+///   a preset in Typst. This is because it is owned by MathWorks and is not public.
```
Nit:

```suggestion
///   a preset in Typst because it is owned by MathWorks and is not public.
```

laurmaedje commented on crates/typst/src/export/pdf/gradient.rs:
```diff
@@ -0,0 +1,199 @@
+        let (first, second) = (window[0], window[1]);
+
+        // Skip stops with the same position.
+        if first.offset.unwrap().get() == second.offset.unwrap().get() {
```
This was marked resolved without any change or comment. Did I miss something?

Dherse replied:
I must have glossed over it, and yes we can just unwrap, when building the list of gradients (`process_stops`), the result **always** contains a `Ratio` therefore they can always be unwrapped safely.

laurmaedje replied:
In that case, could we not store an option in the gradient in the first place? By having a separate stop type for the public interface and the gradient itself.

Dherse replied:
I mean the gradient actually stores it unwrapped, I'll change it so that it uses (also non-allocating) and internal method ;)

Dherse commented:
For the first one I have a good idea of what causes it, and the other one is me only giving it to points, and the two end points are the same color 😅

laurmaedje commented on crates/typst/src/export/pdf/mod.rs:
```diff
@@ -3,11 +3,13 @@
 mod page;
 
 pub use self::color::{ColorEncode, ColorSpaces};
+use self::gradient::PdfGradient;
```
This was also marked resolved even though it is still like this. :/

laurmaedje commented:
I just pushed a commit with some tidying up. While doing that, I noticed the following things:

- `gradient.linear(dir: 70deg)` feels wrong to me. And `gradient.dir() -> angle`, too. Maybe we can have `angle` as the main argument and accessor and `dir` is a separate argument that internally sets the angle? Kinda like `set page(paper: _)` is a shorthand for `width` and `height`.

- I also think the name `relative` is not clear as it's an adjective while what we're looking for is really a noun describing an object. Perhaps there is a better name (`base`, `scope`, `container` is the direction I'm thinking although I'm not a huge fan of either of those).

- I wonder whether the new `Angle::quadrant` method and the `Quadrant` enum should be exposed? I haven't seen real motivation for this public-facing addition.

- In gradient.rs, you have the line `let (sinh, cosh) = h.sin_cos();`. Isn't sinh different from sin?

- The bezier bbox logic that was lifted from GitHub could be replaced by a call to `kurbo`, which is already in our dependency tree. (Would need to add it to `typst`'s Cargo.toml, it's only in `typst-library` right now). The less copy-pasted complex math logic in the code base, the better, in my opinion.

Dherse commented:
> gradient.linear(dir: 70deg) feels wrong to me. And gradient.dir() -> angle, too. Maybe we can have angle as the main argument and accessor and dir is a separate argument that internally sets the angle? Kinda like set page(paper: _) is a shorthand for width and height.

I'll do that in a bit, first I'm finishing some fixes for HSV and HSL interpolation

> I also think the name relative is not clear as it's an adjective while what we're looking for is really a noun describing an object.

Imo it's fine since it's trying to describe what the coordinate system is relative to. And I agree that the other names are not much better. `relative-to` is also a bit ugly imo.

> I wonder whether the new Angle::quadrant method and the Quadrant enum should be exposed? I haven't seen real motivation for this public-facing addition.

I don't mind either way but since I have a lot of operations depending on quadrants (especially with the other types and on text) it's needed at least internally. I can make it private if you want 🤷‍♂️ 

> In gradient.rs, you have the line let (sinh, cosh) = h.sin_cos();. Isn't sinh different from sin?

The code is correct (doubled checked it like 10 mins ago), it's sin(h) and cos(h) I admit the names are a bit problematic, it will be changed in my commit with corrections for HSL and HSV

> The bezier bbox logic that was lifted from GitHub could be replaced by a call to kurbo, which is already in our dependency tree. (Would need to add it to typst's Cargo.toml, it's only in typst-library right now). The less copy-pasted complex math logic in the code base, the better, in my opinion.

I didn't know, will change.

Dherse commented:
@laurmaedje @reknih I think I have covered everything you mentioned, apart from:

- Presets are still arrays
- `relative` is still called `relative`

Dherse commented:
As discussed on Discord, I changed the way color maps are handled so that they're purely arrays of colors and no longer use functions with a # of stop param
